### PR TITLE
[Sofa.Type] Clean Quat

### PR DESCRIPTION
**Pull request by damienmarchal**
- moves code from header to .inl
- use typedef to express the inner types instead of specifying them all the time
  The inner types are: Quat, Vec3, Mat3x3, Real
- add Quaternion to forward declaration file.
- remove the use of the implicit automatically converting functions(BREAKING) 
eg: template<Real2> void toMatrix(Matrix<3,3,Real2>){}
As there is no more hidden type conversion the caller's have to handle themselves
the type conversion before calling instead of relying on the inner-conversion. This is
good as it propagate to the user-level that there is type conversion involved instead of
hidden that fact. The consequence is that it is not possible to do stuff like that anymore
Quat<float>::toMatrix(Mat<3,3,double)>

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 1]
```diff
@@ -92,74 +94,36 @@ class SOFA_TYPE_API Quat
...
     {
-        _q[0]=0.0;
-        _q[1]=0.0;
-        _q[2]=0.0;
-        _q[3]=1.0;
+        set(0.0,0.0,0.0,1);
     }
 
-    void fromFrame(type::Vec<3,Real>& x, type::Vec<3,Real>&y, type::Vec<3,Real>&z);
+    void fromFrame(Vec3& x, Vec3&y, Vec3&z);
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    void fromFrame(const Vec3& x, const Vec3&y, const Vec3&z);
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 2]
```diff
@@ -280,32 +269,29 @@ void Quat<Real>::normalize()
 }
 
 template<class Real>
-void Quat<Real>::fromFrame(type::Vec<3,Real>& x, type::Vec<3,Real>&y, type::Vec<3,Real>&z)
+void Quat<Real>::fromFrame(Vec3& x, Vec3&y, Vec3&z)
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
void Quat<Real>::fromFrame(const Vec3& x, const Vec3&y, const Vec3&z)
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 3]
```diff
@@ -92,74 +94,36 @@ class SOFA_TYPE_API Quat
...
+    auto operator/(const Real &r) const -> Quat;
     void operator*=(const Real &r);
     void operator/=(const Real &r);
 
     /// Given two Quats, multiply them together to get a third quaternion.
 
-    Quat quatVectMult(const type::Vec<3,Real>& vect);
-
-    Quat vectQuatMult(const type::Vec<3,Real>& vect);
+    auto quatVectMult(const Vec3& vect) -> Quat;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    auto quatVectMult(const Vec3& vect) const -> Quat;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 4]
```diff
@@ -92,74 +94,36 @@ class SOFA_TYPE_API Quat
...
+    auto vectQuatMult(const Vec3& vect) -> Quat;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    auto vectQuatMult(const Vec3& vect) const -> Quat;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 5]
```diff
@@ -189,36 +182,32 @@ void Quat<Real>::operator/=(const Real& r)
 
 
 template<class Real>
-Quat<Real> Quat<Real>::quatVectMult(const type::Vec<3,Real>& vect)
+auto Quat<Real>::quatVectMult(const Vec3& vect) -> Quat
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
auto Quat<Real>::quatVectMult(const Vec3& vect) const -> Quat
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 6]
```diff
@@ -189,36 +182,32 @@ void Quat<Real>::operator/=(const Real& r)
...
+    ret._q[0] = _q[3] * vect[0] + _q[1] * vect[2] - _q[2] * vect[1];
+    ret._q[1] = _q[3] * vect[1] + _q[2] * vect[0] - _q[0] * vect[2];
+    ret._q[2] = _q[3] * vect[2] + _q[0] * vect[1] - _q[1] * vect[0];
 
     return ret;
 }
 
 template<class Real>
-Quat<Real> Quat<Real>::vectQuatMult(const type::Vec<3,Real>& vect)
+auto Quat<Real>::vectQuatMult(const Vec3& vect) -> Quat
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
auto Quat<Real>::vectQuatMult(const Vec3& vect) const -> Quat
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 7]
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
...
-            quat[2] = c1 * c2 * s3 - s1 * s2 * c3;
-            quat[3] = c1 * c2 * c3 + s1 * s2 * s3;
-            break;
-        }
-
-        Quat quatResult{ quat[0], quat[1], quat[2], quat[3] };
-        return quatResult;
-    }
-
+    static auto createQuaterFromEuler( Vec3 v, EulerOrder order = EulerOrder::ZYX) -> Quat;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    static auto createQuaterFromEuler(const Vec3& v, EulerOrder order = EulerOrder::ZYX) -> Quat;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 8]
```diff
@@ -801,4 +729,206 @@ void Quat<Real>::operator*=(const Quat<Real>& q1)
...
+        return Quat(0,0,0,1);
+
+    Real nor = 1/phi;
+    Real s = Real(sin(phi/2));
+    return Quat( a[0]*s*nor, a[1]*s*nor,a[2]*s*nor, Real(cos(phi/2)));
+}
+
+
+template<class Real>
+auto Quat<Real>::createQuaterFromEuler( Vec3 v, EulerOrder order) -> Quat
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
auto Quat<Real>::createQuaterFromEuler(const Vec3& v, EulerOrder order) -> Quat
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 9]
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
...
-            a[0] = -a[0];
-            a[1] = -a[1];
-            a[2] = -a[2];
-            a[3] = -a[3];
-        }
-
-        Quat q = b.inverse() * a;
-        return q;
-    }
+    auto quatDiff( Quat a, const Quat& b) -> Quat;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    static auto quatDiff( Quat a, const Quat& b) -> Quat;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 10]
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
...
 
     /// Return the eulerian vector resulting of the movement between 2 quaternions
-    type::Vec<3,Real> angularDisplacement( Quat a, const Quat& b)
-    {
-        return quatDiff(a,b).quatToRotationVector();    // Use of quatToRotationVector instead of toEulerVector:
-                                                        // this is done to keep the old behavior (before the
-                                                        // correction of the toEulerVector function).
-    }
+    auto angularDisplacement( Quat a, const Quat& b) -> Vec3;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    static auto angularDisplacement( const Quat& a, const Quat& b) -> Vec3;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 11]
```diff
@@ -801,4 +729,206 @@ void Quat<Real>::operator*=(const Quat<Real>& q1)
...
+        a[3] = -a[3];
+    }
+
+    Quat q = b.inverse() * a;
+    return q;
+}
+
+/// Return the eulerian vector resulting of the movement between 2 quaternions
+template<class Real>
+auto Quat<Real>::angularDisplacement(Quat a, const Quat& b) -> Vec3
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
auto Quat<Real>::angularDisplacement(const Quat& a, const Quat& b) -> Vec3
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 12]
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
...
-
+    void setFromUnitVectors(const Vec3& vFrom, const Vec3& vTo);
 
     // Print the quaternion (C style)
+    [[deprecated("The function print will be removed soon")]]
     void print();
-    Quat<Real> slerp(Quat<Real> &q1, Real t);
-    Quat<Real> slerp2(Quat<Real> &q1, Real t);
+
+    auto slerp(Quat &q1, Real t) -> Quat;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    auto slerp(const Quat &q1, Real t) const -> Quat;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h [hunk 13]
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
...
+    auto slerp2(Quat &q1, Real t) -> Quat;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h]:
```suggestion
    auto slerp2(const Quat &q1, Real t) const-> Quat;
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 14]
```diff
@@ -655,10 +591,10 @@ Quat<Real> Quat<Real>::slerp(Quat<Real> &q1, Real t)
 
 // Given an axis and angle, compute quaternion.
 template<class Real>
-Quat<Real> Quat<Real>::slerp2(Quat<Real> &q1, Real t)
+auto Quat<Real>::slerp2(Quat &q1, Real t) -> Quat
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
auto Quat<Real>::slerp2(const Quat &q1, Real t) const -> Quat
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 15]
```diff
@@ -620,26 +564,18 @@ void Quat<Real>::slerp(const Quat& a, const Quat& b, Real t, bool allowFlip)
...
-//template<class Real>
-//    std::ostream& operator<<(std::ostream& out, Quat<Real> Q)
-//{
-//	return (out << "(" << Q._q[0] << "," << Q._q[1] << "," << Q._q[2] << ","
-//				<< Q._q[3] << ")");
-//}
-
 template<class Real>
-Quat<Real> Quat<Real>::slerp(Quat<Real> &q1, Real t)
+auto Quat<Real>::slerp(Quat &q1, Real t) -> Quat
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]:
```suggestion
auto Quat<Real>::slerp(const Quat &q1, Real t) const -> Quat
```

**Review by alxbilger — COMMENTED** [review]
Minor suggestions related to constness and references

**Comment by damienmarchal** [comment]
Thanks for the suggestion @alxbilger, 

Did you use a tool to make the suggestion on github or did you do that manually ? It is the first time I try the "commit in batch" feature from git. In general we are more often making PR from a suggestion-branch to the PR branch.

**Comment by alxbilger** [comment]
> Thanks for the suggestion @alxbilger,
> 
> Did you use a tool to make the suggestion or did you do that manually ? It is the first time I try the "commit in batch" feature from git. In general we are more often making PR from a suggestion-branch to the PR branch.

I used GitHub only. It's also the first time I see that we can directly suggest things in a PR review. GitHub notified me about this new feature when I tried to comment a line the first time. I am not sure it's a good practice, because I did not compile my suggestions at all.

File: SofaKernel/modules/Sofa.Type/CMakeLists.txt [hunk 16]
```diff
@@ -48,6 +48,7 @@ sofa_find_package(Sofa.Config REQUIRED)
 add_library(${PROJECT_NAME} SHARED ${HEADER_FILES} ${SOURCE_FILES})
 
 target_link_libraries(${PROJECT_NAME} PUBLIC Sofa.Config)
+target_compile_definitions(${PROJECT_NAME} PRIVATE "-D SOFA_BUILD_SOFA_TYPE")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/CMakeLists.txt]:
```suggestion
target_link_libraries(${PROJECT_NAME} PUBLIC Sofa.Config)
```

File: SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl [hunk 17]
```diff
@@ -801,4 +729,206 @@ void Quat<Real>::operator*=(const Quat<Real>& q1)
...
+    m[0][2] = (2 * (_q[2] * _q[0] + _q[1] * _q[3]));
+
+    m[1][0] = (2 * (_q[0] * _q[1] + _q[2] * _q[3]));
+    m[1][1] = (1 - 2 * (_q[2] * _q[2] + _q[0] * _q[0]));
+    m[1][2] = (2 * (_q[1] * _q[2] - _q[0] * _q[3]));
+
+    m[2][0] = (2 * (_q[2] * _q[0] - _q[1] * _q[3]));
+    m[2][1] = (2 * (_q[1] * _q[2] + _q[0] * _q[3]));
+    m[2][2] = (1 - 2 * (_q[1] * _q[1] + _q[0] * _q[0]));
+}
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl]: @Younesssss remark: why m[3] is not handled here?

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

**Comment by damienmarchal** [comment]
There is PR in SofaPython3 to keep synchronized with this PR. 
https://github.com/sofa-framework/SofaPython3/pull/133

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

**Pull request merged**